src/NHibernate/Engine/Query/HQLQueryPlan.cs

```diff
@@ -93,60 +94,52 @@ public void PerformList(QueryParameters queryParameters, ISessionImplementor ses
 
 			bool hasLimit = queryParameters.RowSelection != null && queryParameters.RowSelection.DefinesLimits;
 			bool needsLimit = hasLimit && Translators.Length > 1;
-			QueryParameters queryParametersToUse;
+
+			IList combinedResults = results ?? new List<object>();
+
+			IEnumerable<object> items = GetResultSets(queryParameters, session, needsLimit).SelectMany(l => l.Cast<object>());
+
 			if (needsLimit)
 			{
 				Log.Warn("firstResult/maxResults specified on polymorphic query; applying in memory!");
-				RowSelection selection = new RowSelection();
-				selection.FetchSize = queryParameters.RowSelection.FetchSize;
-				selection.Timeout = queryParameters.RowSelection.Timeout;
-				queryParametersToUse = queryParameters.CreateCopyUsing(selection);
+				if (queryParameters.RowSelection.FirstRow != RowSelection.NoValue)
+					items = items.Skip(queryParameters.RowSelection.FirstRow);
+
+				if (queryParameters.RowSelection.MaxRows != RowSelection.NoValue)
+					items = items.Take(queryParameters.RowSelection.MaxRows);
 			}
-			else
+
+			ArrayHelper.AddAll(combinedResults, items);
+		}
+
+		private QueryParameters GetModifiedQueryParametersForInMemoryPaging(QueryParameters originalParameters, int alreadyTraversedRecords)
+		{
+			RowSelection selection = new RowSelection();
+			selection.FetchSize = originalParameters.RowSelection.FetchSize;
+			selection.Timeout = originalParameters.RowSelection.Timeout;
+			if (originalParameters.RowSelection.MaxRows != RowSelection.NoValue)
 			{
-				queryParametersToUse = queryParameters;
+				selection.MaxRows = Math.Max(0, (originalParameters.RowSelection.FirstRow == RowSelection.NoValue ? 0 : originalParameters.RowSelection.FirstRow) 
+												+ originalParameters.RowSelection.MaxRows
+												- alreadyTraversedRecords);
```

Comment:
You cannot just assume that the results returned by the first transformer will go before the results of the second transformer.